signalfx-java/src/main/java/com/signalfx/signalflow/WebSocketTransport.java

```diff
@@ -0,0 +1,592 @@
+        transportConnection.sendMessage(request);
+    }
+
+    private static Map<String, String> buildParams(Object... params) {
+        Preconditions.checkArgument(params.length % 2 == 0);
+        ImmutableMap.Builder<String, String> builder = new ImmutableMap.Builder<String, String>();
+        for (int i = 0; i < params.length; i += 2) {
+            if (params[i] != null && params[i + 1] != null) {
+                builder.put(params[i].toString(), params[i + 1].toString());
+            }
+        }
+        return builder.build();
+    }
+
+    /**
+     * Builder of WebSocket Transport Instance
+     */
+    public static class TransportBuilder {
+
+        private String token;
+        private String protocol = "wss";
+        private String host = SignalFlowTransport.DEFAULT_HOST;
+        private int port = 443;
+        private int timeout = 1;
+        private int version = 2;
+
+        public TransportBuilder(String token) {
+            this.token = token;
+        }
+
+        public TransportBuilder setProtocol(String protocol) {
+            this.protocol = protocol;
+            return this;
+        }
+
+        public TransportBuilder setHost(String host) {
+            this.host = host;
+            return this;
+        }
+
+        public TransportBuilder setPort(int port) {
+            this.port = port;
+            return this;
+        }
+
+        public TransportBuilder setTimeout(int timeout) {
+            this.timeout = timeout;
+            return this;
+        }
+
+        public TransportBuilder setAPIVersion(int version) {
+            this.version = version;
+            return this;
+        }
+
+        public WebSocketTransport build() {
+            SignalFxEndpoint endpoint = new SignalFxEndpoint(this.protocol, this.host, this.port);
+            WebSocketTransport transport = new WebSocketTransport(this.token, endpoint,
+                    this.version, this.timeout);
+            return transport;
+        }
+    }
+
+    /**
+     * Special type of StreamMessage for conveying websocket/connection errors to channels
+     */
+    protected static class SignalFlowExceptionStreamMessage extends StreamMessage {
+
+        protected SignalFlowException exception;
+
+        public SignalFlowExceptionStreamMessage(final SignalFlowException exception) {
+            super("error", null, exception.getMessage());
+            this.exception = exception;
+        }
+
+        public SignalFlowException getException() {
+            return this.exception;
+        }
+    }
+
+    /**
+     * Value Object that handles binary data message conversion in construction
+     */
+    protected static class TransportDataMessage {
+
+        protected byte version;
+        protected byte messageType;
+        protected String channelName;
+        protected long logicalTimestampMs;
+        protected ArrayList<Map<String, Object>> data = new ArrayList<Map<String, Object>>();
```

Comment:
Always use interface types in the variable definition. No need to specify the generic type in the variable assignment, either:

```java
protected List<Map<String, Object>> data = new ArrayList<>();
```